### ISSUE TITLE
Test parametrization does not affect the fixture dependency tree

### ISSUE THREAD

**Issue by egnartsms** [pytest-dev/pytest#2220]
pytest v3.0.5

```python
@pytest.fixture
def fix1(fix2):
   return fix2 + '1'

@pytest.fixture
def fix2(fix3):
   return fix3 + '2'

@pytest.fixture
def fix3():
   print("fix3 instantiated!!")
   return '3'

@pytest.mark.parametrize('fix2', ['2'])
def test_it(fix1):
   assert fix1 == '21'
```

This example demonstrates that _fix3_ is still instantiated, although _fix2_ - _fix3_ dependency should have been broken because of parametrization.

**Comment by nicoddemus** [comment]
Thanks for the report! 👍

**Comment by egnartsms** [comment]
Can I expect this to be fixed by some date? Is this issue included in any future release?

**Comment by The-Compiler** [comment]
Nobody is currently working on this - so unless you submit a fix yourself (or get a support contract with [Merlinux](http://merlinux.eu/), there isn't really a guarantee things will get fixed at some point. Keep in mind we're all volunteers :wink:

**Comment by timofurrer** [comment]
What is the expected behavior for the above example? 
I might be able to work on it.

**Comment by nicoddemus** [comment]
@timofurrer 

Here's a more complete example:

```python
import pytest

@pytest.fixture
def fix1(fix2):
   return fix2 + '1'

@pytest.fixture
def fix2(fix3):
    print('fix2 instantiated!!')
    return fix3 + '2'

@pytest.fixture
def fix3():
    print("fix3 instantiated!!")
    return '3'

@pytest.mark.parametrize('fix2', ['2'])
def test_it(fix1):
   assert fix1 == '21'
```

Running `pytest -s`:

```
fix3 instantiated!!
.
1 passed in 0.01 seconds
```

As can be seen, even though `fix2` was not instantiated because it was shadowed by the parametrize call in `test_it`, `fix3` was still instantiated because of the `fix2` dependency to `fix3`.

At least that's how I understood it, @egnartsms feel free to correct me.

**Comment by timofurrer** [comment]
`fix1` is instantiated, too. Which of the fixtures should be instantiated? Non of them?

**Comment by nicoddemus** [comment]
@timofurrer yeah it is a little confusing at first, here's a summary of what's happening:

* **fixture** `fix1` should be instantiated because `test_it` explicitly asks for it in its parameter list;
* **parameter** `fix2` is injected into `test_it` because of the parametrize call;
* **fixture** `fix2` is never instantiated because it is not part of the parameter list of `test_it`;
* **fixture** `fix3` *should* never be instantiated because, while it is a dependency of **fixture** `fix2`, `fix2` itself was not instantiated and so shouldn't `fix3`.

It seems **fixture** `fix2` is at some point being considered a dependency of `test_it` (which implicitly also makes `fix3` a dependency), and then later `fix2` is replaced by the **parameter** `fix2` but `fix3` is kept. It warrants more investigation for sure. 😅

**Comment by nicoddemus** [comment]
Fixed by #3629, thanks @egnartsms!
